File: src/main/java/com/bleudev/nine_lifes/custom/effect/AmethysmEffect.java
package com.bleudev.nine_lifes.custom.effect;

import com.bleudev.nine_lifes.custom.CustomDamageTypes;
import com.bleudev.nine_lifes.custom.CustomEffects;
import com.bleudev.nine_lifes.networking.payloads.StartAmethysmScreenEffectPayload;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.attribute.AttributeContainer;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.entity.effect.StatusEffectCategory;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;

import java.util.Objects;

public class AmethysmEffect extends StatusEffect {
    private int ticks;
    private LivingEntity entity;

    public AmethysmEffect() {
        super(StatusEffectCategory.NEUTRAL, 0xff00ff);
    }

    @Override
    public boolean canApplyUpdateEffect(int duration, int amplifier) {
        return true;
    }

    @Override
    public boolean applyUpdateEffect(ServerWorld world, LivingEntity entity, int amplifier) {
        if (!world.isClient()) {
            ticks++;
            this.entity = entity;
            if (!entity.isSleeping()) {
                if (ticks % 20 == 0)
                    entity.damage(world, CustomDamageTypes.of(world, CustomDamageTypes.AMETHYSM_DAMAGE_TYPE), 1f);
                entity.setVelocity(0, 0.1, 0);
                entity.velocityModified = true;
            }
            entity.setGlowing(true);
        }
        return true;
    }

    @Override
    public void onApplied(LivingEntity entity, int amplifier) {
        super.onApplied(entity, amplifier);

        ticks = 0;
        this.entity = entity;

        if (entity instanceof ServerPlayerEntity player) try {
            ServerPlayNetworking.send(player, new StartAmethysmScreenEffectPayload(Objects.requireNonNull(player.getStatusEffect(CustomEffects.AMETHYSM)).getDuration()));
        } catch (NullPointerException ignored) {}
    }

    @Override
    public void onRemoved(AttributeContainer attributeContainer) {
        super.onRemoved(attributeContainer);
        if (!entity.hasStatusEffect(StatusEffects.GLOWING))
            entity.setGlowing(false);
    }
}

File: 26.1/src/main/kotlin/com/bleudev/nine_lifes/custom/effect/mob/AmethysmEffect.kt
package com.bleudev.nine_lifes.custom.effect.mob

import com.bleudev.nine_lifes.custom.NineLifesDamageSources
import com.bleudev.nine_lifes.custom.NineLifesMobEffects
import com.bleudev.nine_lifes.custom.packet.payload.StartAmethysmScreen
import com.bleudev.nine_lifes.util.sendPacket
import net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.effect.MobEffectCategory
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.ai.attributes.AttributeMap

class AmethysmEffect : MobEffect(MobEffectCategory.NEUTRAL, 0xff00ff) {
    private var ticks = 0
    private var entity: LivingEntity? = null

    override fun shouldApplyEffectTickThisTick(duration: Int, amplifier: Int): Boolean = true

    override fun applyEffectTick(level: ServerLevel, entity: LivingEntity, amplifier: Int): Boolean {
        if (!level.isClientSide) {
            ticks++
            this.entity = entity
            if (!entity.isSleeping) {
                if (ticks % 20 == 0)
                    entity.hurtServer(level, NineLifesDamageSources.amethysm(level), 1f)
                entity.setDeltaMovement(.0, 0.05, .0)
                entity.needsSync = true

                if (entity is ServerPlayer)
                    entity.connection.send(ClientboundSetEntityMotionPacket(entity))
            }
            entity.setGlowingTag(true)
        }
        return true
    }

    override fun onEffectAdded(entity: LivingEntity, amplifier: Int) {
        super.onEffectAdded(entity, amplifier)
        ticks = 0
        this.entity = entity
        if (entity is ServerPlayer)
            entity.sendPacket(StartAmethysmScreen(entity.getEffect(NineLifesMobEffects.AMETHYSM)?.duration ?: return))
    }

    override fun removeAttributeModifiers(attributeMap: AttributeMap) {
        super.removeAttributeModifiers(attributeMap)
        (this.entity as LivingEntity).let {
            if (!it.hasEffect(MobEffects.GLOWING)) it.setGlowingTag(false)
        }
    }
}